### DISCUSSION TITLE
Boolean operators overhaul

### DISCUSSION THREAD

**Discussion by mathieucloutier86** [dnGrep/dnGrep#646] [Ideas]
Extract from the Boolean operators 

"When using multiple Boolean operators, they must be the same: you cannot use both ANDs and ORs in one search.
The AND and OR operators are not case sensitive, and must be separated from the search pattern by a single space.
Additional spaces become part of the search pattern."

Boolean operators are most powerful when combined. I wondered if it was considered making it possible to use  both AND, OR , NOT in the same search.

Best,
Mathieu

**Comment by doug24** [comment]
Yes, I agree. I made those restrictions to greatly simplify the implementation.  I'll have to see how difficult it would be to make these changes.

Combining AND, OR, and NOT means there will operator precedence, so we will also have to add parentheses to the syntax to override the operator precedence.  Adding parentheses for the Boolean operators complicates parsing with included expressions, it will likely require a formal parser to extract the sub-expressions and get the order of evaluation.

**Comment by doug24** [comment]
I have been thinking about this and writing some code to try things out.  I still think the challenge is tokenizing/parsing the input strings, particularly to handle the patterns of regular expressions.  There needs to be some way to delimit a regular expression pattern if it contains the Boolean key words or parentheses.  I don't think there are any characters on the keyboard that aren't already used in regular expressions. The back tick is rarely used in regular expressions, but is not on keyboards in all languages.  I think the next best option is angle brackets, but they conflict with named capture groups.  I have defined both to delimit the patterns or to preserve whitespace in the pattern as in these examples:
```
`((\"".+?\"")|('.+?'))` AND `\d+`
<((\"".+?\"")|('.+?'))> OR \d+
`sand ` or ` floor`
<sand > and < floor>
sand and  floor
```
(the last example would not include additional whitespace, despite the extra space in the expression)
Combining operators:
```
a and b or c
a and (b or c)
a and not b
(a and b) or (c or d)
```

A couple of other thoughts:
- The Boolean key words are not case sensitive, and the tokenizer distinguishes whole words (so `sand` won't be confused with `and`).
- The Boolean expression is evaluated in the search range, which is normally one line of text in a file.  But if the `Multiline` option is enabled, the search range is the whole file.
- The patterns evaluate to True if a pattern is matched in the search range, and False if not found.  Introducing the NOT operator means you can have a True result with no matching text.  I don't know if dnGrep can handle this since it is built on returning a set of matches, not a collection of lines that have no matches. I haven't gotten far enough to try this yet.
- It would be just as easy to also add NAND and NOR in addition to AND, OR, NOT and parentheses.  I do wonder if mixing these operators and the operator precedence will be confusing when constructing expressions.
- Each patten in the Boolean expression is searched separately, which can be expensive for long expressions. It would be good to be able to short-circuit the search if the expression reaches the point it cannot return True.  I'm not sure how much work that would be.

**Comment by mathieucloutier86** [comment]
Thank you for your work on this. I'd like to be more helpful, alas, I'm a
librarian, not a programmer. Should you decide to go ahead and code
something alongside the lines you've described, just let me know, I'll try
my best to assist you, if that can help you.

Best,
Mathieu

**Comment by mathieucloutier86** [comment]
Hi Doug,

I haven't had much time to try the boolean system that you have implemented
back in March, but I did try it tonight. I'm very much please with the
result so far. I'm writing to you because I'd like to contribute a small
sum to encourage your work on the software. What would be the best way to
do that ?

Kind regards,
Mathieu

**Reply by doug24** [reply]
Thanks/you're welcome!  I haven't set anything up for payments so far, but have thought about it.  I'll let you know if I do.

**Comment by vatterspun** [comment]
I'm happy to give this a separate issue thread but if we're looking at operators, I'd love to also get a "near" operator.  That is sometimes you know there are two words that were closely associated in a file but not as a keyword like "cats do not love dogs" could be sorted by "cats NEAR dogs."  Lucene does this via "cats dogs"~10 (as described here: http://www.waldoavenuerestorationbranch.org/docs/RSPhelp/DocFetcher/help/English/DocFetcher_Manual_files/Query_Syntax.html)

I realize there are some options inside of regex with the lookahead / lookbehind options, but they're fairly complex.  Hoping for something a bit smoother.

**Comment by doug24** [comment]
I think you are asking for a huge amount of work, so I'm going to say no to this one.  I don't have the time to take on a big project.

Lucerne is an indexing library and builds a database of the contents of the files/data that it has been asked to index.  Using the generated index Lucerne can perform complex queries very quickly.   But that is well outside of scope of grep application.

**Reply by vatterspun** [reply]
No problem -- was more focused on the "near" bit and less on Lucene support, but I hear you.
